### PR TITLE
Signup: add FSE to /start flow

### PR DESCRIPTION
**Pull request by creativecoder**
#### Changes proposed in this Pull Request

Adds initial support for FSE to the /start flow.

- Uses an assignment function for easily changing the percentage of users that have new sites created as FSE sites
- Adds an additional callback to the design picker step that determines if FSE is active, once the theme is set up
- Redirects to the Site editor when the signup flow is complete (unless the "Write" intent is selected)
- Logs a tracks event `calypso_fse_enrolled` when a user creates a site and property site_enrolled: true/false, depending on if the users/site is enrolled in FSE.

#### Testing instructions

* Visit /start and create a site
* At the intent step, select "Build"
* In the Design picker, select a design that is based on Blockbase, such as Quadrat, Geologist, Zoologist, etc
* You should be redirected to the Site Editor after site setup is complete
* You should see FSE beta themes in Appearance themes

Note that if FSE isn't active for any reason (like the current theme doesn't support it), you won't be directed to the Site editor after setup, **but** you still should see the FSE treatment of the themes showcase because the site is eligible for FSE.

Related to #60255

### CODE REVIEW

File: client/lib/signup/step-actions/index.js [hunk 1]
```diff
@@ -39,6 +43,22 @@ import { getSiteId } from 'calypso/state/sites/selectors';
...
+		return false;
+	}
+
+	// Force the FSE Beta on development and Horizon environments.
+	if ( config.isEnabled( 'full-site-editing/beta-opt-in' ) ) {
+		return true;
+	}
+
+	// Check if the user ID is part of the eligible percentage.
+	return userId % 100 < fse_beta_rollout_percentage_start_flow;
```
Comment (p-jackson) [inline client/lib/signup/step-actions/index.js]: Given there isn't an opt-in screen to deal with, should we use a real explat A/B test for the allocation? Benefits are we could adjust the allocation in the explat interface and could do a real comparison of the two cohorts.

paYE8P-1oZ-p2#comment-1301

File: client/lib/signup/step-actions/index.js [hunk 2]
```diff
@@ -377,8 +399,14 @@ export function setDesignOnSite( callback, { siteSlug, selectedDesign } ) {
...
-		.then( () => {
-			callback();
+		.then( () =>
+			wpcom.req.get( {
+				path: `/sites/${ siteSlug }/block-editor`,
+				apiNamespace: 'wpcom/v2',
+			} )
+		)
+		.then( ( data ) => {
+			callback( null, { isFSEActive: data?.is_fse_active ?? false } );
```
Comment (creativecoder) [inline client/lib/signup/step-actions/index.js:414]: This API call for `is_fse_active` detects if
- The site is eligible for the FSE beta
- The Site editor is turned on for the site (which should be the case for all FSE eligible sites)
- And that the current theme supports FSE

File: client/signup/config/flows.js [hunk 3]
```diff
@@ -109,6 +109,10 @@ function getDestinationFromIntent( dependencies ) {
 		return `/post/${ siteSlug }`;
 	}
 
+	if ( isFSEActive ) {
+		return `/site-editor/${ dependencies.siteSlug }`;
+	}
+
```
Comment (creativecoder) [inline client/signup/config/flows.js]: This will redirect to the Site editor after the site-setup flow... I'm not sure if there are other flows where we'd want to do this.

Comment (p-jackson) [reply client/signup/config/flows.js]: All the site creation flows (e.g. `onboarding`, `free`, etc.) eventually redirect to the `site-setup` flow, so this is the only one we have to worry about 👍

File: client/lib/signup/step-actions/index.js (hunk 2 above)
Comment (p-jackson) [reply client/lib/signup/step-actions/index.js:414]: Great! I wasn't sure how we were going to figure out whether the selected theme was an FSE theme. This works, but do you think the theme API should be modified so that it returns for each theme whether it's FSE? That way it'd work more like Gutenboarding, where the hardcoded list of available-designs has an is_fse flag.

File: client/signup/config/flows.js [hunk 4]
```diff
@@ -109,6 +109,10 @@ function getDestinationFromIntent( dependencies ) {
 		return `/post/${ siteSlug }`;
 	}
 
+	if ( isFSEActive ) {
```
Comment (p-jackson) [inline client/signup/config/flows.js]: I think there's a slight logic hole here where the user can choose the "Write" intent, ask to "skip to my home" and then end up being redirected to the site editor. If the user is in the test cohort then maybe it should change this to a "skip to site editor" button. But in the mean time should it be this:

```suggestion
	if ( isFSEActive && intent !== 'write' ) {
```

**Review by p-jackson — COMMENTED** [review]
> (unless the "Write" intent is selected)

Great, this answers one of the questions I had.

File: client/lib/signup/step-actions/index.js (hunk 1 above)
Comment (creativecoder) [reply client/lib/signup/step-actions/index.js]: We didn't in the past, for the reasons outlined here: paYE8P-12q-p2

My understanding is that Explat is not a good fit for gradually ramping up the number of users into a new experience, like we plan to do here.

File: client/lib/signup/step-actions/index.js (hunk 2 above)
Comment (creativecoder) [reply client/lib/signup/step-actions/index.js:414]: I don't think we need to modify the API, as we already have the `block-templates` tag, which indicates the theme supports the Site editor (it's used for both universal and block-only themes).

File: client/lib/signup/step-actions/index.js (hunk 1 above)
Comment (p-jackson) [reply client/lib/signup/step-actions/index.js]: Thanks, wasn't aware of some of those limitations. I've followed up on Marko's suggestion to consider it paYE8P-1oZ-p2#comment-1323

File: client/lib/signup/step-actions/index.js [hunk 5]
```diff
@@ -39,6 +43,22 @@ import { getSiteId } from 'calypso/state/sites/selectors';
 const Visibility = Site.Visibility;
 const debug = debugFactory( 'calypso:signup:step-actions' );
 
+const isUserAssignedFSEBeta = ( userId ) => {
+	const fse_beta_rollout_percentage_start_flow = 10;
```
Comment (zaguiini) [inline client/lib/signup/step-actions/index.js]: Can we extract that constant to the global scope and upper case its name? Conventions 😬

File: client/lib/signup/step-actions/index.js [hunk 6]
```diff
@@ -39,6 +43,22 @@ import { getSiteId } from 'calypso/state/sites/selectors';
...
+
+	if ( ! userId ) {
+		return false;
+	}
```
Comment (zaguiini) [inline client/lib/signup/step-actions/index.js:51]: Will this ever be the case? In what cases will someone that is not logged in be able to create a site?

File: client/lib/signup/step-actions/index.js (hunk 2 above)
Comment (zaguiini) [reply client/lib/signup/step-actions/index.js:414]: Can't we modify the `sites/new` response to include the information that we fetch from `sites/%s/block-editor` for us, so we can avoid that extra request?

**Review by zaguiini — APPROVED** [review]
Works great 👍

File: client/lib/signup/step-actions/index.js (hunk 2 above)
Comment (p-jackson) [reply client/lib/signup/step-actions/index.js:414]: The theme is actually being selected some time after the `sites/new` endpoint has been called. In the current incarnation of the flow the site is always created using the Zoologist theme, and then later changed after the user has seen the design picker.

You still make a valid point (we could modify `themes/mine` or `sites/%s/theme-setup` to return `is_fse_actve`) but just wanted to clarify in case there was confusion.

File: client/lib/signup/step-actions/index.js (hunk 6 above)
Comment (p-jackson) [reply client/lib/signup/step-actions/index.js:51]: No, not given the way onboarding currently works. We know this because we happen to know the function is only called from within `createSiteWithCart()`. But I'm imagining `isUserAssignedFSEBeta` is a useful lib function that may be pulled out into a common place before this whole FSE rollout project is through. So I think it's better to be defensive.

Comment (p-jackson) [reply client/lib/signup/step-actions/index.js:51]: Actually, this is _can_ happen. If you test the flow with a brand new user (in incognito mode for example) then `userId === null`

**Review by p-jackson — CHANGES_REQUESTED** [review]
Found an issue in the case where the user has just signed up for the first time

- Test in incognito mode
- Test in `calypso.localhost:3000` or an environment where the `full-site-editing/beta-opt-in` flag is `true`
- Go to `/start` and create a new user
- Watch the network when you click **start with a free site**

Notice that the `sites/new` endpoint is called with `enable_fse: false`

This is happening because `userId === null`

The signup framework runs in an interesting mode when a new user has just been created. The page hasn't refreshed yet which means the cookie isn't set and most of the Calypso code doesn't recognise that there is a current user (hence why `getCurrentUserId()` returns `null`). Instead there's a `bearer_token` dependency that is used to make authenticated requests, like the one that creates the new site.

I'm looking now for where the user ID is made available within the signup framework ...

**Comment by p-jackson** [comment]
I tried adding the ID of brand new users to signup dependencies so that it can be accessed during site creation. This diff seemed to do the trick for me.

https://gist.github.com/p-jackson/27699b9360bedd26614d32067659aa32

Edit: There are a couple of other steps that look like they would need their `providesDependencies` updated, but those steps are actually unused. #60378 Cleans that up.

File: client/lib/signup/step-actions/index.js (hunk 6 above)
Comment (vindl) [reply client/lib/signup/step-actions/index.js:51]: > If you test the flow with a brand new user (in incognito mode for example) then userId === null

This rollout is primarily intended for brand new users, so we should change this to make sure they are included too. If there are no better options, instead of `userId` we could fall back to a random number between 0 and 99.

Comment (p-jackson) [reply client/lib/signup/step-actions/index.js:51]: Ideally, we'd have something deterministic because we want to change other parts of their experience based on their eligibility i.e. only showing non-universal block themes to eligible users. We could use a random number here and then use the `/block-editor` endpoint whenever we wanted to check their availability later on. But my idea here https://github.com/Automattic/wp-calypso/pull/60330#issuecomment-1019666501 makes the ID of brand new users available in the required place.

File: client/signup/config/steps-pure.js [hunk 7]
```diff
@@ -160,13 +161,15 @@ export function generateSteps( {
...
 				'allowUnauthenticated',
+				'user_id',
 			],
 			optionalDependencies: [
 				'bearer_token',
 				'username',
 				'marketing_price_group',
 				'plans_reorder_abtest_variation',
 				'allowUnauthenticated',
+				'user_id',
```
Comment (creativecoder) [inline client/signup/config/steps-pure.js:172]: @p-jackson I've applied your patch, thank you! I was wondering why user_id is specified as an optional dependency in the `user-new` flow, but not in the others?

**Comment by vindl** [comment]
One thing that we'll need to add to this PR is a tracks event for FSE enrolled sites. Currently, we are keeping track of enrolled sites with `calypso_fse_enrolled` event that's dispatched at the beginning of the `/new` flow.

We can either do the same thing here or move the event after the last signup step when it's guaranteed that FSE site will be created.

**Comment by creativecoder** [comment]
> One thing that we'll need to add to this PR is a tracks event for FSE enrolled sites. Currently, we are keeping track of enrolled sites with calypso_fse_enrolled event that's dispatched at the beginning of the /new flow.
>
> We can either do the same thing here or move the event after the last signup step when it's guaranteed that FSE site will be created.

@vindl Done in https://github.com/Automattic/wp-calypso/pull/60330/commits/1c63c824e06e136eb6ee3503943a0c4cbdadd17f

**Comment by creativecoder** [comment]
@p-jackson Thanks for the patch... that seems to work fine in my testing when creating a new user as part of the flow. Just one question I asked in code comments.

Also, I wasn't able to test the sign up with Google/Apple flows explicitly in a local environment, might have to double check it works as expected in production, unless you know of an easy way to test those flows locally.

File: client/signup/config/steps-pure.js (hunk 7 above)
Comment (p-jackson) [reply client/signup/config/steps-pure.js:172]: This step looks like it was added for this project pau2Xa-1hW-p2
tl;dr it looks like the plan was to allow users through the flow without signing up, hence why `bearer_token` and `username` are optional.

The thread says the project's on hold, and it's possible we may be able to remove this whole flow. But in the meantime, I think it's fair for `user_id` to be optional whenever `username` is optional.

Comment (p-jackson) [reply client/signup/config/steps-pure.js:172]: I just tested `/start/onboarding-registrationless` on this branch (which is the only flow that uses this step), and it behaves exactly the same as it does in production.

File: client/lib/signup/step-actions/index.js [hunk 8]
```diff
@@ -286,6 +308,10 @@ export function createSiteWithCart( callback, dependencies, stepData, reduxStore
 		return;
 	}
 
+	if ( newSiteParams.options.enable_fse ) {
+		recordTracksEvent( 'calypso_fse_enrolled', { flow: flowToCheck } );
```
Comment (p-jackson) [inline client/lib/signup/step-actions/index.js]: We analyse tracks events on a per-user basis, but this event gets recorded once per-site. Has that created any confusion in the `/new` rollout?
For instance
- user creates a site
- their ID % 100 is > 10 so they're not enrolled
- we don't record tracks event
- user creates another site at a later date
- their ID % 100 is > 50 (we've now rolled out FSE to more users) so they are enrolled
- we do record tracks event

When analysing how this user interacts with their first site, we might incorrectly think they're an FSE user because they have recorded a `calypso_fse_enrolled` event at _some_ point.

Maybe we've limited our analyse to users who only have a single site, so this hasn't been a problem.

Comment (p-jackson) [reply client/lib/signup/step-actions/index.js]: I see that it was added to `/new` in #60328 and it just fires as soon as you enter the flow, regardless of whether the user successfully creates a site. So I see it's just a rough guide, so probably don't need to overthink it here.

**Review by p-jackson — APPROVED** [review]
> I wasn't able to test the sign up with Google/Apple flows explicitly in a local environment, might have to double check it works as expected in production, unless you know of an easy way to test those flows locally.

Unfortunately not, I think I've had to just wait till it was on `staging` in the past.

This change is working as expected for me. Tested both as an existing user and a brand new user. Tried "write" and "build" flows, and selecting block and classic themes.

The design picker isn't showing the "block-only" themes like Twenty Twenty-Two, but if it's fine to ship that separately ([I think it is](https://github.com/Automattic/wp-calypso/issues/60255#issuecomment-1020784076)) then this is good to go.

File: client/lib/signup/step-actions/index.js (hunk 8 above)
Comment (vindl) [reply client/lib/signup/step-actions/index.js]: > When analysing how this user interacts with their first site, we might incorrectly think they're an FSE user because they have recorded a calypso_fse_enrolled event at some point.

That's true, we haven't made efforts to differentiate these in a better way, but I'd imagine it's not a large fraction of users/sites. I'm open to better ideas for how to deal with this in a more precise way. My intuition is that similar effects are in play for non-enrolled users, so on average they shouldn't bias the stats one way or the other. 🤔

Comment (vindl) [inline client/lib/signup/step-actions/index.js]: @creativecoder I have one more change request related to stats after discussing this with @kurt213.

Could we add a `site_enrolled: true` prop for enrolled users, and also record `calypso_fse_enrolled` with `site_enrolled: false` prop for non-enrolled users? It's just for the purposes of easier segmentation in the Tracks Funnels.

And would it be possible to record this event when the last step in the flow is completed? (currently, it's happening between plans selection and design selection steps) We are doing something similar for `/new` in https://github.com/Automattic/wp-calypso/pull/60450.

**Comment by creativecoder** [comment]
PR now updated

- Triggers `calypso_fse_enrolled` for all new sites, adding the `site_enrolled` property as true/false, depending on fse enrollment for the user.
- Show only block-based theme (universal and block-only) in the design picker step for FSE enrolled users/sites

File: client/lib/signup/step-actions/index.js (hunk 8 above)
Comment (creativecoder) [reply client/lib/signup/step-actions/index.js]: I've added the `site_enrolled` property and the `calypso_fse_enrolled` event is now triggered for all new sites created in /start, specifying site_enrolled: true/false depending on FSE enrollment.

I'm hesitant to move the event trigger out of the site creation step. The signup flow is intentionally modular, and technically that design step could be triggered by any flow, at any point, for any existing site, including multiple times for the same site. I think keeping the event tied to site creation gives us better parity with the /new flow, and prevents some confusion in limiting the event to being triggered once per site.

Comment (p-jackson) [reply client/lib/signup/step-actions/index.js]: The most likely way the user won't follow the `site creation -> site design` pathway is by going through the checkout.

1. Create a site at `/start`
2. Choose a non-free plan
3. Go to checkout
4. The user bails out of the usual flow e.g. empties their checkout basket, follows a link shown during an upsell, etc.

In these situations, the user doesn't land in the `site-setup` flow.

So in this situation would we like `calypso_fse_enrolled` to be recorded or not? I think we should because the user will have the default theme (which is Zoologist, a universal theme) and will have access to the site editor if they're in the test cohort. So really they _are_ enrolled.
So I think we should keep `calypso_fse_enrolled` where it currently is in the flow.

File: client/signup/steps/design-picker/index.jsx [hunk 9]
```diff
@@ -59,8 +62,19 @@ export default function DesignPickerStep( props ) {
 	const [ selectedDesign, setSelectedDesign ] = useState( null );
 	const scrollTop = useRef( 0 );
 
+	// Limit themes to those that support the Site editor, if site is fse eligible
+	const siteId = useSelector( ( state ) => getSiteId( state, dependencies.siteSlug ) );
```
Comment (p-jackson) [inline client/signup/steps/design-picker/index.jsx:66]: Because we're now `dependencies.siteSlug` directly, I realised we need to test a whole other path. When the user adds a custom domain during signup and goes through checkout then we don't use the `?siteSlug` query param, we use the `?siteId` query param. We have code that ensures the `signupDependencies.siteSlug` is correctly populated, and it appears to be working. But I wanted to mention it as an FYI.
If this turns out to cause a problem then the `getSelectedSiteId()` selector is another option

**Review by p-jackson — COMMENTED** [review]
The `/themes` request doesn't wait for the result of the `/block-editor` which causes a duplicate request. That's because when `/block-editor` returns and we find out they're enrolled in FSE it's sending off _another_ request to `/themes` with the updated filter param.

To fix the duplicated request we could get the `isLoading` or `isSuccess` flag returned by `useBlockEditorSettingsQuery()` and pass it to the `useThemeDesignQuery()`'s `enabled` parameter?

This change is also firing off an unnecessary request to `/block-editor` in the `/start/do-it-for-me` flow. Not a big deal because the result just gets ignored. But we could disable the `useBlockEditorSettingsQuery()` hook if `props.useDIFMThemes` is true.

I also noticed that every time the user navigates away from and back to the design picker we fire another request to `/block-editor`. Outside the scope of this PR, but it seems like the `useBlockEditorSettingsQuery()` hook isn't caching the result correctly.

File: client/lib/signup/step-actions/index.js (hunk 2 above)
Comment (zaguiini) [reply client/lib/signup/step-actions/index.js:414]: @creativecoder any thoughts on the API response modification? ☝️

File: client/lib/signup/step-actions/index.js (hunk 8 above)
Comment (vindl) [reply client/lib/signup/step-actions/index.js]: Makes sense, let's keep it as is then.

**Review by vindl — APPROVED** [review]

**Comment by creativecoder** [comment]
@p-jackson Thanks for these great suggestions preventing duplicate API requests. I've implemented theme in https://github.com/Automattic/wp-calypso/pull/60330/commits/972874cd7ec432fb6d90ba943ed7aa93ce898c59 and all appears to be working

- The themes request is called once, after block editor settings have loaded
- The do-it-for-me flow skips the block editor settings request

File: client/signup/steps/design-picker/index.jsx (hunk 9 above)
Comment (creativecoder) [reply client/signup/steps/design-picker/index.jsx:66]: Thanks for pointing this out. I just tested this with a new user, including a plan and domain purchase. I see the siteId param in the url when I get to the site setup flow and it the design picker loads the correct FSE themes without any problems that I can see.
